cmd/shipperctl/cmd/decommission.go

```diff
@@ -28,221 +34,205 @@ var (
-	counter := 0
-
-	configurator, err := configurator.NewClusterConfiguratorFromKubeConfig(kubeConfigFile, managementClusterContext)
+func collectReleases(kubeClient kubernetes.Interface, shipperClient shipperclientset.Interface) ([]ReleaseAndFilteredAnnotations, error) {
+	var updateAnnotations []ReleaseAndFilteredAnnotations
+	namespaceList, err := kubeClient.CoreV1().Namespaces().List(metav1.ListOptions{})
 	if err != nil {
-		return err
-	}
-
-	namespaceList, err := configurator.KubeClient.CoreV1().Namespaces().List(metav1.ListOptions{})
-	if err != nil {
-		return err
+		return nil, err
 	}
 	var errList []string
 	for _, ns := range namespaceList.Items {
-		releaseList, err := configurator.ShipperClient.ShipperV1alpha1().Releases(ns.Name).List(metav1.ListOptions{})
+		releaseList, err := shipperClient.ShipperV1alpha1().Releases(ns.Name).List(metav1.ListOptions{})
 		if err != nil {
 			errList = append(errList, err.Error())
 			continue
 		}
 		for _, rel := range releaseList.Items {
-			trueClusters := getFilteredSelectedClusters(&rel)
-			if len(trueClusters) == 0 {
-				counter++
-			}
-		}
-	}
+			selectedClusters := releaseutil.GetSelectedClusters(&rel)
+			filteredClusters := cmdutil.FilterSelectedClusters(releaseutil.GetSelectedClusters(&rel), clusters)
+			if len(filteredClusters) > 0 {
+				sort.Strings(filteredClusters)
 
-	cmd.Println("Number of *releases* that are scheduled only on decommissioned clusters: ", counter)
-	if len(errList) > 0 {
-		return fmt.Errorf(strings.Join(errList, ","))
-	}
-	return nil
-}
+				filteredClusterAnnotation := strings.Join(filteredClusters, ",")
+				if filteredClusterAnnotation == rel.Annotations[shipper.ReleaseClustersAnnotation] {
+					continue
+				}
 
-func getFilteredSelectedClusters(rel *shipper.Release) []string {
-	clusters := releaseutil.GetSelectedClusters(rel)
-	var trueClusters []string
-	for _, cluster := range clusters {
-		if !filters.SliceContainsString(decommissionedClusters, cluster) {
-			trueClusters = append(trueClusters, cluster)
+				updateAnnotations = append(
+					updateAnnotations,
+					ReleaseAndFilteredAnnotations{
+						OldClusterAnnotation:      strings.Join(selectedClusters, ","),
+						FilteredClusterAnnotation: filteredClusterAnnotation,
+						Namespace:                 rel.Namespace,
+						Name:                      rel.Name,
+					})
+				continue
+			}
+			isContender, err := cmdutil.IsContender(&rel, shipperClient)
+			if err != nil {
+				errList = append(errList, err.Error())
+				continue
+			}
+			isIncumbent, err := cmdutil.IsIncumbent(&rel, shipperClient)
+			if err != nil {
+				errList = append(errList, err.Error())
+				continue
+			}
+			// emptying the scheduled clusters annotation will make shipper re-schedule this release.
+			// If this release is a history release, Shipper will enforce strategy correctly (no capacity and no traffic)
+			// However, if this release is contender, Shipper will possibly give capacity and traffic to a release
```

Comment:
This behavior of shipper rescheduling the contender? yes I have
If the cluster annotation is empty shipper will choose clusters again (function `chooseClusters` in `release_controller.go` line 574)

Reply:
Great, than I would suggest be more explicit in the comment: 

`Shipper will possibly give capacity and traffic to a release...`
^ this sentence seems to assume that we don't really know what's going to happen.

What do you think ? Have I understood correctly ?

Reply:
`possibly` because it depends on some things:
If this release *should* get capacity and traffic (because it is a contender, or it is an incumbent during a rollout) and currently is scheduled *only* on decommissioned clusters. This means this release is currently *not* active, and by deleting the cluster annotations shipper will "reactivate" it. 

This is the scenario I tried to explain in this comment, and we want to avoid it because it creates side affects without the owner of this application being aware of it.

Reply:
Aaah I see, thanks. The sentence is clear now 💯